### ISSUE TITLE
Call to removeOlderResults when the discovery service is stopped

### ISSUE THREAD

**Issue by lolodomo** [eclipse/smarthome#6213]
I see a lot of bindings removing the inbox entries when the discovery service is stopped.

For example, the Hue binding: https://github.com/eclipse/smarthome/blob/7176f8961a4caba501fe38bab52a44c8705011c3/extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/discovery/HueLightDiscoveryService.java#L83

If I am not wrong, this code is removing the inbox entries when openHAB is shutdown ?

Is it the expected behaviour ?

**Comment by lolodomo** [comment]
Ok it can make sense.
Now I understand why my hue lights are discovered at each restart.
But in case this is what is expected, I think it should be handled by the framework, not by each binding ?
The problem with that is that the information "mark as ignored" is lost.
For that last reason, I don't thing it is a good idea to remove disxovered things.

**Comment by kaikreuzer** [comment]
> But in case this is what is expected, I think it should be handled by the framework, not by each binding ?

No, because only the binding knows, whether it is easy to "re-discover" upon a bundle start or not.
Think of radio devices that report their existence only every hour, you wouldn't want to remove those inbox entries. While with stuff that you can find through UPnP, you can make sure that the results are recent and it does not make sense to keep old stuff that does not exist anymore.

> The problem with that is that the information "mark as ignored" is lost.

Ok, that's an issue that should be dealt with indeed.

**Comment by maggu2810** [comment]
I am not really sure if we should remove the discovered things / the inbox entries at all.

If a bridge is removed, we don't remove the things that uses that bridge.
If a binding is removed, we don't remove the things of that binding.

Why is this different for inbox entries?

We can argue that discovered stuff is volatile.
But if the discovery is "complicated" we can also argue that it should be kept also if a binding is temporary removed.

WRT https://github.com/eclipse/smarthome/issues/4302 there could be a REST method "remove all inbox entries that match filter" and a filter "binding ID == ..." or "bridge UID == ...". Then a UI that e.g. uninstalls a binding or delete a bridge can provide a message dialog if all inbox entries should be cleaned up or not.

**Comment by kaikreuzer** [comment]
> But if the discovery is "complicated" we can also argue that it should be kept also if a binding is temporary removed.

This is the status quo, yes.
But if it is easy, there is no point in keeping results that cannot be approved at all.
My personal experience is that there a lot of outdated entries in the Inbox, which are simply annoying and the bindings should much better deal with the removal of outdated stuff.

> Why is this different for inbox entries?

Things are entities that are defined & managed by the user. Inbox results are not.

**Comment by lolodomo** [comment]
If the criteria is easy discovery, this must be added to several bindings.

**Comment by lolodomo** [comment]
I could propose to add an option for the method `removeOlderResults` to avoid removing the entries that are marked as ignored. We could use it at binding shutdown. This will avoid a new discovery without the ignored mark at restart of the binding.

**Comment by lolodomo** [comment]
The problem is that we cannot make a difference between a binding uninstall and a binding shutdown.

**Comment by kaikreuzer** [comment]
Maybe the inbox should keep (=persist) a list of ignored UIDs separately of the results, so that results stay hidden no matter how often they come and go?

**Comment by lolodomo** [comment]
And we will not remove this list of ignored UIDs at service shutdown.
Ok that could be a solution.
But that would mean that we could have ignored UID with no corresponding discovery result. We should at least provide a way to delete them through a console command.
I will try to implement this solution taking care to not break the REST API and the behaviour in Paper UI.

**Comment by lolodomo** [comment]
In fact, there is a simple solution consisting in adding a new `removeOlderResults` method with a boolean parameter used to request keeping or not the ignored discovery results.
Thsi method will be used by bindings when deactivating the discovery service to keep ignored discovery results.

**Comment by kaikreuzer** [comment]
I am not sure if this is a valid solution. If you ignore a UPnP device and this briefly disappears and reappears on your network, you would still want to have it ignored, don't you?
